flink-optimizer/src/main/java/org/apache/flink/optimizer/dag/OuterJoinNode.java

```diff
@@ -0,0 +1,139 @@
+
+	private List<OperatorDescriptorDual> dataProperties;
+
+	/**
+	 * Creates a new two input node for the optimizer plan, representing the given operator.
+	 *
+	 * @param operator The operator that the optimizer DAG node should represent.
+	 */
+	public OuterJoinNode(OuterJoinOperatorBase<?, ?, ?, ?> operator) {
+		super(operator);
+
+		this.dataProperties = getDataProperties();
+	}
+
+	private List<OperatorDescriptorDual> getDataProperties() {
+		OuterJoinOperatorBase<?, ?, ?, ?> operator = getOperator();
+
+		OuterJoinType type = operator.getOuterJoinType();
+
+		JoinHint joinHint = operator.getJoinHint();
+		joinHint = joinHint == null ? JoinHint.OPTIMIZER_CHOOSES : joinHint;
+
+		List<OperatorDescriptorDual> list = new ArrayList<>();
+		switch (joinHint) {
+			case OPTIMIZER_CHOOSES:
+				list.add(getSortMergeDescriptor(type, true));
+				list.add(getSortMergeDescriptor(type, false));
+				break;
+			case REPARTITION_SORT_MERGE:
+				list.add(getSortMergeDescriptor(type, false));
+				break;
+			case REPARTITION_HASH_FIRST:
+			case REPARTITION_HASH_SECOND:
+			case BROADCAST_HASH_FIRST:
+			case BROADCAST_HASH_SECOND:
+			default:
+				throw new CompilerException("Invalid join hint: " + joinHint + " for outer join type: " + type);
+		}
+
+		Partitioner<?> customPartitioner = operator.getCustomPartitioner();
+		if (customPartitioner != null) {
+			for (OperatorDescriptorDual desc : list) {
+				((AbstractJoinDescriptor) desc).setCustomPartitioner(customPartitioner);
+			}
+		}
+		return list;
+	}
+
+	private OperatorDescriptorDual getSortMergeDescriptor(OuterJoinType type, boolean broadcastAllowed) {
+		if (type == OuterJoinType.FULL) {
+			return new SortMergeFullOuterJoinDescriptor(this.keys1, this.keys2);
+		} else if (type == OuterJoinType.LEFT) {
+			return new SortMergeLeftOuterJoinDescriptor(this.keys1, this.keys2, broadcastAllowed);
+		} else {
+			return new SortMergeRightOuterJoinDescriptor(this.keys1, this.keys2, broadcastAllowed);
+		}
+	}
+
+	@Override
+	public OuterJoinOperatorBase<?, ?, ?, ?> getOperator() {
+		return (OuterJoinOperatorBase<?, ?, ?, ?>) super.getOperator();
+	}
+
+	@Override
+	protected List<OperatorDescriptorDual> getPossibleProperties() {
+		return dataProperties;
+	}
+
+	@Override
+	public String getOperatorName() {
+		return "Outer Join";
+	}
+
+	@Override
+	protected void computeOperatorSpecificDefaultEstimates(DataStatistics statistics) {
+		OuterJoinType type = getOperator().getOuterJoinType();
+
+		long card1 = getFirstPredecessorNode().getEstimatedNumRecords();
+		long card2 = getSecondPredecessorNode().getEstimatedNumRecords();
+
+		if (card1 < 0 || card2 < 0) {
+			this.estimatedNumRecords = -1;
+		} else {
+			if (type == OuterJoinType.LEFT) {
+				this.estimatedNumRecords = card1;
+			} else if (type == OuterJoinType.RIGHT) {
+				this.estimatedNumRecords = card2;
+			} else if (type == OuterJoinType.FULL) {
+				this.estimatedNumRecords = Math.max(card1, card2);
```

Comment:
Should be `Math.max(card1, card2)` for all outer join types. An outer join returns at least as many records as an inner join and `Math.max(card1, card2)` is the estimate of the inner join (see JoinNode).